lib/packets/packetElements/clientVersion.ts

```diff
@@ -0,0 +1,31 @@
+export class ClientVersion {
+  constructor(
+    public year: number,
```

Comment:
I'm really not a fan of having constructor args declared as `public` instead of them being specified on the class itself and assigned in the constructor.

```js
export class ClientVersion {
  year: number,
  month: number,
  day: number,
  revision: number

  constructor (year: number, month: number, day: number, revision: number) {
    this.year = year
    ...
  }
}

Reply:
I completely disagree with this. Why do you prefer this way?

Reply:
I was thinking of just making the properties private with a public getter to prevent modification

Reply:
> I completely disagree with this. Why do you prefer this way?

It is in my opinion counter-intuitive, and also not the [recommended way](https://www.typescriptlang.org/docs/handbook/classes.html) as you now have to parse the constructor to look for types, rather than just having to look at the defined class variables. If the constructor was not _empty_ then maybe it would feel less bad, but at the moment we're just defining constructors to define arguments..

Reply:
I disagree that it's counterintuitive, there are other languages which have support for this. PHP just added it, for example.

Reply:
Fair enough, I did say _in my opinion_. Just wanted to point out that it's not the way that 99% of open source JS is written, and may confuse people. When in doubt, go with the status quo, but if you all really like having empty constructors then go for it :)

Reply:
I prefer it because I think it's much cleaner and easier to read as you're only defining a property once instead of three times.

Reply:
I don't *like* having an empty constructor, I just think it's less bloated than repeating each variable name 4 times, and each type twice

Reply:
And FWIW just because one practice is the common practice doesn't mean everybody should follow it.

Reply:
```js
interface ClientVersionConfig {
  year: number,
  month: number,
  day: number,
  revision: number
}

export class ClientVersion {
  config: ClientVersionConfig

  constructor (config: ClientVersionConfig) {
    this.config = config
  }
}
```

If your main flag is repetition, then we can always do something like this.

Reply:
> And FWIW just because one practice is the common practice doesn't mean everybody should follow it. I don't mean to sound rude, I'm just for avoiding herd mentality.

No, I agree, but if you want to deviate from said common practice there **should** be an extremely good reason for it, which imo "duplicating types twice" is not. But like I said, if you all want to keep doing public var declarations in constructors then go for it

Reply:
It doesn't make sense for a ClientVersion class to have a config, `year`, `month`, `day` and `revision` are all inherent properties to ClientVersion

Reply:
Then we have to call `.config` as a prefix to all property accessors too

Reply:
I also can't find anything that says using parameter properties is bad.

Reply:
> I also can't find anything that says using parameter properties is bad.

I never said they were bad, they do make sense in some specific scenarios, but _zero_ of the examples in the typescript docs use them, and I think you'll struggle to find any high profile TS project that does either

Reply:
How about we ditch classes, like we've all discussed 

```ts

interface ClientVersion {
  year: number,
  month: number,
  day: number,
  revision: number,

  serialise(): number,
  matches(otherVersion: ClientVersion): boolean
}

export function parseClientVersion (version: number): ClientVersion {
  let year = Math.floor(version / 25000);
  version %= 25000;
  let month = Math.floor(version / 1800);
  version %= 1800;
  let day = Math.floor(version / 50);
  let revision = version % 50;

  return clientVersion(year, month, day, revision)
}

export function clientVersion (
  year: number,
  month: number,
  day: number,
  revision: number
): ClientVersion {
  const serialise = (): number => {
    return (
      this.year * 25000 + this.month * 1800 + this.day * 50 + this.revision
    );
  }

  const matches = (otherVersion: ClientVersion): boolean {
    return (
      this.year == otherVersion.year &&
      this.month == otherVersion.month &&
      this.day == otherVersion.day &&
      this.revision == otherVersion.revision
    );
  }
  
  return {
    year,
    month,
    day,
    revision,
    serialise,
    matches
  }
}
```

Reply:
I mean the link you posted above *literally* does what we're doing:

```ts
class Octopus {
  readonly numberOfLegs: number = 8;
  constructor(readonly name: string) {}
}

let dad = new Octopus("Man with the 8 strong legs");
dad.name;
```

Reply:
> I mean the link you posted above _literally_ does what we're doing:
> 
> ```ts
> class Octopus {
>   readonly numberOfLegs: number = 8;
>   constructor(readonly name: string) {}
> }
> 
> let dad = new Octopus("Man with the 8 strong legs");
> dad.name;
> ```

Yep fair enough I missed that one.

Reply:
From personal use, having class members defined both in the class and in the constructor has just made it much harder to build a mental model of the class, and how to access it. I think it's an antipattern.